File: py/1208.py
from sys import stdin, stdout
from collections import defaultdict


def dfs(graph, n, v, match, visited) -> bool:
    for i in graph[v]:
        if not visited[i]:
            visited[i] = True
            if not match[i] or dfs(graph, n, match[i], match, visited):
                match[i] = v
                return True
    return False


def main() -> None:
    input = stdin.readline
    print = stdout.write
    while True:
        try:
            n, m = input().split()
            n = int(n)
            m = int(m)
            gr = defaultdict(lambda: defaultdict(bool))
            for i in range(m):
                p, f = input().split()
                p = int(p)
                f = int(f)
                gr[p][f] = True
            res = 0
            match = defaultdict(bool)
            for i in gr:
                res += dfs(gr, n, i, match, defaultdict(bool))
            print(str(n - res) + '\n')
        except ValueError:
            break


if __name__ == '__main__':
    main()

File: cpp/1208.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, m;
vector<int> graph[1001], visited, match;

int dfs(int v)
{
    if (visited[v]) {
        return 0;
    }
    visited[v] = 1;
    for (int i = 0; i < graph[v].size(); i++) {
        int r = graph[v][i];
        if (match[r] == -1 || dfs(match[r])) {
            match[r] = v;
            return 1;
        }
    }
    return 0;
}

int main()
{
    cin.tie(0)->sync_with_stdio(false);
    int p, f;
    while (cin >> n >> m) {
        for (int i = 0; i < m; i++) {
            cin >> p >> f;
            graph[p].push_back(f + n);
        }
        int res = n;
        match.assign(2 * n + 1, -1);
        for (int i = 1; i <= n; i++) {
            visited.assign(n + 1, 0);
            res -= dfs(i);
        }
        cout << res << '\n';
        for (int i = 1; i <= n; i++) graph[i].clear();
    }
    return 0;
}